tsdb/wal.go

```diff
@@ -566,24 +566,22 @@ func (w *SegmentWAL) cut() error {
 		}
 		// Finish last segment asynchronously to not block the WAL moving along
 		// in the new segment.
-		go func() {
-			w.actorc <- func() error {
-				off, err := hf.Seek(0, io.SeekCurrent)
-				if err != nil {
-					return errors.Wrapf(err, "finish old segment %s", hf.Name())
-				}
-				if err := hf.Truncate(off); err != nil {
-					return errors.Wrapf(err, "finish old segment %s", hf.Name())
-				}
-				if err := hf.Sync(); err != nil {
-					return errors.Wrapf(err, "finish old segment %s", hf.Name())
-				}
-				if err := hf.Close(); err != nil {
-					return errors.Wrapf(err, "finish old segment %s", hf.Name())
-				}
-				return nil
+		w.actorc <- func() error {
```

Comment:
Ok, I think this will probably solve leak on close, but also this code can block easily. Think about the case when `w.actorc` is full.
 

To know for sure we might want to run some benchmarks. I think it's not the best idea to kill performance potentially only to fix close leak (when we close just once per Prometheus life) ):